feedback

pkumar999 opened the pull request.

TheCodeMaster210 committed:
Added microservices section

Krackido committed:
adding mechanic endpoints

Krackido committed:
adding md documentation and swagger test file

DylanB1205 committed:
Update init_shop_flow_pro.sql

DylanB1205 committed:
Add files via upload

DylanB1205 committed:
Add files via upload

DylanB1205 committed:
Add files via upload

DylanB1205 committed:
Add files via upload

DylanB1205 committed:
Add files via upload

DylanB1205 committed:
Add files via upload

DylanB1205 committed:
Delete MicroServices/shopflowproservice directory

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

Krackido committed:
adding endpoint README

Krackido committed:
Merge branch 'main' of github.com:WSU-kduncan/cs3900-autobody-4 into main

changes

TheCodeMaster210 committed:
Update Services-README.md

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Mechanic.java:
```diff
@@ -0,0 +1,39 @@
+@Getter
+@Setter
+@Builder
+@Table(name = "Mechanic")
```
table & column names must be in lower case and separated by underscore if have more than one word  (e.g. mechanic). Otherwise you have to define JPA naming strategy to recognize other naming.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Mechanic.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Id
+    @Column(name = "mechanic_id")
+    private Integer mechanicId;
```
Many places you're considering mechanic id as String and also defined by admin when creating mechanic profile. I would like to have String in entity model class & DTO and update table to support VARCHAR with correct column name.
Also you are using MID as column name in database table but that's function in DB used for extract a specific portion of a text string

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Mechanic.java:
```diff
@@ -0,0 +1,39 @@
+    @Column(name = "mechanic_first_name", nullable = false)
+    private String firstName;
+
+    @Column(name = "mechanic_last_name", nullable = false)
```
update your DB script to have correct column name as given here

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceController.java:
```diff
@@ -0,0 +1,31 @@
+package shopflowproservice.controller;
```
wrong package name here, it should be com.wsu.shopflowproservice.controller;

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceController.java:
```diff
@@ -0,0 +1,31 @@
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import static com.wsu.workorderproservice.utilities.Constants.MESSAGE;
```
you have incorrect package name here and above imports. It should be shopflowproservice instead of workorderproservice

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceController.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Map;
+
+@RestController
+@RequestMapping("/service")
```
REST API design required to have plural form of resource URI. Change "/service" to "/services"

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceController.java:
```diff
@@ -0,0 +1,31 @@
+    private final ServiceService serviceService;
+
+    @GetMapping
+    public ResponseEntity<ServiceResponseDTO> getServices(@RequestParam(required = false) boolean isActive) {
```
you don't have active status column in Service entity model class. So, you don't need to pass unnecessary query parameter isActive here

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/MechanicController.java:
```diff
@@ -0,0 +1,91 @@
+     **/
+    @PostMapping
+    public ResponseEntity<ServiceResponseDTO> save(@RequestBody @Valid MechanicDTO mechanicDTO) {
+        if (!StringUtils.hasLength(mechanicDTO.getId())) {
```
In your Mechanic entity model class and other places you are using mechanicId. Either you can change there as id or use mechanicId here and MechanicDTO class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/MechanicController.java:
```diff
@@ -0,0 +1,91 @@
+     * @return - HTTP status with the updated mechanic DTO
+     **/
+    @PutMapping("/{mechanicId}")
+    public ResponseEntity<ServiceResponseDTO> update(@PathVariable Integer mechanicId,
```
As you are considering mechanicId as String. You must have to use String instead of Integer for mechanicId here.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceOrderController.java:
```diff
@@ -0,0 +1,75 @@
+     * @param serviceOrderId - id used to get serviceORder details
+     * @return - ServiceResponseDTO which contains details of serviceOrder
+     */
+    @GetMapping(value = "{serviceOrderId}")
```
**General Advise:** 
You have GET endpoint to get Service Order by serviceOrderId that's okay for details purpose. But you need another GET endpoint to get Service Orders in paginated response that should allow to filter. Which can help to see specific service orders for the customer or vehicle or specific Service type.

DylanB1205 replied:
So would this be another GET method within the ServiceOrderController with similar components just passing different variables?

pkumar999 replied:
We given example to have summary endpoint in new Controller class because joining many tables and getting summarized data in response for work order. 
If you want to achieve same then you can have in Separate controller & service class for better readability, otherwise you can add new GET endpoint in same controller class also and use JPA method to return ServiceOrder paginated response by applying filter if any.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceOrderController.java:
```diff
@@ -0,0 +1,75 @@
+     * @return - ServiceResponseDTO which include serviceOrder saved entity
+     */
+    @PostMapping
+    public ResponseEntity<ServiceResponseDTO> addServiceOrder(@RequestBody ServiceOrderDTO serviceOrder) {
```
**General Advise:** 
Reference name should be based on class name. Here it should be **serviceOrderDTO** instead of **serviceOrder**.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceOrderController.java:
```diff
@@ -0,0 +1,75 @@
+    @PutMapping(value = "{serviceOrderId}")
+    public ResponseEntity<ServiceResponseDTO> updateServiceOrder(@PathVariable Integer serviceOrderId, @RequestBody ServiceOrderDTO serviceOrder) {
+        return new ResponseEntity<>(ServiceResponseDTO.builder().meta(Map.of(MESSAGE, "ServiceOrder updated successfully"))
+        .data(ServiceOrderService.update(serviceOrderId, serviceOrder)).build(), HttpStatus.OK);
```
use reference instead of class name to call update() method of ServiceOrderService. 
e.g. serviceOrderService.update()

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/controller/ServiceOrderController.java:
```diff
@@ -0,0 +1,75 @@
+    @PostMapping
+    public ResponseEntity<ServiceResponseDTO> addServiceOrder(@RequestBody ServiceOrderDTO serviceOrder) {
+        return new ResponseEntity<>(ServiceResponseDTO.builder().meta(Map.of(MESSAGE, "ServiceOrder added successfully."))
+        .data(ServiceOrderService.add(serviceOrder)).build(), HttpStatus.CREATED);
```
use reference instead of class name to call add() method of ServiceOrderService.
e.g. serviceOrderService.add()

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/dto/MechanicDTO.java:
```diff
@@ -0,0 +1,28 @@
+@Builder
+public class MechanicDTO {
+
+    private Integer id;
```
most places you are using this field name as mechanicId and datatype as String. I'd like to change here also datatype as String and field name as mechanicId. 
String can give flexibility to store mechanic id if that's created based on any logic (e.g. first character of first name and full last name).

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/dto/ServiceDTO.java:
```diff
@@ -0,0 +1,18 @@
+public class ServiceDTO {
+    private String serviceName;
+    private int serviceID;
+}
```
**General Advise:**
It's better if you can store service cost also based on service type

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/dto/ServiceOrderDTO.java:
```diff
@@ -0,0 +1,31 @@
+public class ServiceOrderDTO {
+
+    private Integer serviceOrderId;
+    private String VIN;
```
As per Java naming field name should be in CAMEL case (i.e. first character should start in lower case). Here you can have vin instead of VIN.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/dto/ServiceOrderDTO.java:
```diff
@@ -0,0 +1,31 @@
+    private Integer serviceOrderId;
+    private String VIN;
+    private Integer mechanicId;
+    private Integer serviceId;
```
you don't need ServiceId here as you are going to have multiple line items (i.e. if service order include multiple service type e.g. Oil Change and Tire Change). 

you can remove serviceId field from this DTO class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/dto/ServiceOrderDTO.java:
```diff
@@ -0,0 +1,31 @@
+    private Date dateCompleted;
+    private String customerFirstName;
+    private String customerLastName;
+    private Float serviceCost;
```
**General Advise:**
you can have field name totalCost instead of service cost because this cost will be sum of all services cost that's included in this service order.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/dto/ServiceOrderDTO.java:
```diff
@@ -0,0 +1,31 @@
+    private String customerFirstName;
+    private String customerLastName;
+    private Float serviceCost;
+    private Set<ServiceOrderLineItem> lineItems;
```
**General Advise:** 
I don't see any use case to have separate entity for line items in your code. You might be provide list of service type that's include in this service order. 

e.g. `private Set<ServiceType> lineItems; `

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/exception/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,73 @@
+package com.wsu.workorderproservice.exception;
```
You have wrong package name in this class.
Replace **workorderproservice** to **shopflowproservice** for all packages import in this class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/exception/DatabaseErrorException.java:
```diff
@@ -0,0 +1,7 @@
+package com.wsu.workorderproservice.exception;
```
You have wrong package name in this class.
Replace **workorderproservice** to **shopflowproservice** here.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/exception/InvalidRequestException.java:
```diff
@@ -0,0 +1,9 @@
+package com.wsu.workorderproservice.exception;
```
You have wrong package name in this class.
Replace **workorderproservice** to **shopflowproservice** here.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/exception/validation/ValidationError.java:
```diff
@@ -0,0 +1,17 @@
+package com.wsu.workorderproservice.exception.validation;
```
You have wrong package name in this class.
Replace **workorderproservice** to **shopflowproservice** here.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/exception/validation/ValidationErrorResponse.java:
```diff
@@ -0,0 +1,18 @@
+package com.wsu.workorderproservice.exception.validation;
```
You have wrong package name in this class.
Replace **workorderproservice** to **shopflowproservice** here.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Mechanic.java:
```diff
@@ -0,0 +1,39 @@
+    @Column(name = "mechanic_id")
+    private Integer mechanicId;
+
+    @Column(name = "mechanic_first_name", nullable = false)
```
update your DB script to have correct column name as given here

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Service.java:
```diff
@@ -0,0 +1,23 @@
+@Getter
+@Setter
+@Builder
+public class Service {
```
refactor this class name as ServiceType instead of Service because Service is an predefined Interface annotation that's used to declare Service class

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Service.java:
```diff
@@ -0,0 +1,23 @@
+    @Column(name = "ServiceID")
+    private int code;
+    @Column(name = "ServiceName")
+    private String name;
```
**General Advise:**
It's better if you can store service cost also based on service type

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+@Getter
+@Setter
+@Builder
+@Table(name = "ServiceOrder")
```
table name should be in lower case and separated by underscore if it's contains more than two words. e.g. service_order.
Otherwise you have to specify JPA naming strategy for custom naming convention.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "ServiceOrderID")
```
table column name should be in lower case and separated by comma separated if it's contains more than two words. e.g. service_order_id.
Otherwise you have to specify JPA naming strategy for custom naming convention. 

This change need to apply on all column name.
e.g.
vin, mechanic_id, date_received, date_completed, customer_first_name, customer_last_name, total_cost

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+    private Float serviceCost;
+    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
+    @JoinColumn(name = "serviceOrderID", updatable = false)
+    private Set<ServiceOrderLineItem> lineItems;
```
You can use JoinTable to store multiple service type for the service order as you are not storing any additional info there. Because cost can be define in Service Type entity model class.

e.g.
`
@JoinTable(name = "service_order_line_item",
            joinColumns = @JoinColumn(name = "service_order_id"),
            inverseJoinColumns = @JoinColumn(name = "service_id"))
`
`
    @ManyToMany
`
`
    private Set<ServiceType> lineItems;
`

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+    @Column(name = "MID")
+    private Integer mechanicId;
+    @Column(name = "ServiceID")
+    private Integer serviceId;
```
you don't need service id in this entity model class because Service Order can have more than one service type. And it's provided on lineItems.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+    @Column(name = "ServiceOrderID")
+    private Integer serviceOrderId;
+    @Column(name = "VIN")
+    private String VIN;
```
column name & field name should be in lower case e.g. vin

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+    private Integer serviceOrderId;
+    @Column(name = "VIN")
+    private String VIN;
+    @Column(name = "MID")
```
Don't use column name as MID because MID is an DB function that's used to extract substring from string value. 
you can give same column name that's given in mechanic entity model class (i.e. mechanic_id).

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrder.java:
```diff
@@ -0,0 +1,57 @@
+    @Column(name = "CustomerLastName")
+    private String customerLastName;
+    @Column(name = "ServiceCost")
+    private Float serviceCost;
```
**General Feedback:** 
You can have field name as totalCost because you are storing total cost for all line items.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrderLineItemId.java:
```diff
@@ -0,0 +1,16 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class ServiceOrderLineItemId implements Serializable{
```
For JoinTable case you don't need to create Composite Key. You can safely delete this class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrderLineItem.java:
```diff
@@ -0,0 +1,33 @@
+@Getter
+@Setter
+@Builder
+@Table(name = "ServiceOrder")
```
You given wrong table name here because this table used for Service Order. If you want then you can safely delete this class.

**General Advise:**
if you create ManyToMany mapping join table then you don't need to declare JPA entity for service_order_line_item. Safely you can delete this class

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrderLineItem.java:
```diff
@@ -0,0 +1,33 @@
+    @Column(name = "ServiceOrderID")
+    private Integer serviceOrderId;
+    @Id
+    @Column(name = "ServiceCost")
```
you can not have service cost as id.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/ServiceOrderLineItemId.java:
```diff
@@ -0,0 +1,16 @@
+public class ServiceOrderLineItemId implements Serializable{
+
+    private Integer serviceOrderId;
+    
```
**General Advise:**
For composite key second column also need to be specify

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/repository/MechanicRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface MechanicRepository extends JpaRepository<Mechanic, String> {
+
+    @Query(nativeQuery = true, value = "SELECT * FROM Mechanic")
```
Given search filter not applied on SQL and also field mapping not given to apply sorting correctly. 

**SQL should be looks like:**
SELECT mechanic_id as mechanicId, first_name as firstName, last_name as lastName, specialization as specialization FROM mechanic where :search IS NULL OR (first_name = :search OR last_name = :search OR specialization = :search)

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/repository/ServiceOrderRepository.java:
```diff
@@ -0,0 +1,22 @@
+            "INNER JOIN mechanic m on m.MID = s.MID " +
+            "WHERE (:customerLastName IS NULL OR c.CustomerLastName = :customerLastName) " +
+            "AND (:dateCompleted IS NULL OR s.DateCompleted = :dateCompleted)")
+            Page<Object[]> findBySearch(String customerLastName, String VIN, Integer serviceOrderId, Pageable pageable);
```
you were not using this method anywhere and SQL script is incorrect. You can safely either delete or comment this method.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/repository/ServiceOrderRepository.java:
```diff
@@ -0,0 +1,22 @@
+            "AND (:dateCompleted IS NULL OR s.DateCompleted = :dateCompleted)")
+            Page<Object[]> findBySearch(String customerLastName, String VIN, Integer serviceOrderId, Pageable pageable);
+
+            ServiceOrder save(ServiceOrderDTO serviceOrder);
```
you don't need to declare method for saving DTO class. You have to use default save method provided by JPA repository for entity model class.
you can safely delete this method

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/repository/ServiceRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface ServiceRepository extends JpaRepository<Service, String> {
+
+    List<Service> findByActive(boolean active);
```
this method not used anywhere. you can safely delete this method or have additional column active status in ServiceType entity model class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/repository/ServiceRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import java.util.List;
+
+public interface ServiceRepository extends JpaRepository<Service, String> {
```
You have to pass Integer instead of String because you have primary key as Integer.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/MechanicService.java:
```diff
@@ -0,0 +1,129 @@
+        }
+        try {
+            Mechanic mechanic = mapToEntity(mechanicDTO);
+            mechanic.setMechanicId(mechanicDTO.getMechanId());
```
you have typo mistake when getting mechanicId from DTO to set on entity model class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/MechanicService.java:
```diff
@@ -0,0 +1,129 @@
+     * @param mechanicDTO - the details to update
+     * @return - the updated mechanic entity
+     **/
+    public Mechanic update(Integer mechanicId, MechanicDTO mechanicDTO) {
```
Most of places you are using datatype String for mechanic id and it's fit for this case. 
You can change here also String as datatype for mechanic id.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/MechanicService.java:
```diff
@@ -0,0 +1,129 @@
+     * Deletes a mechanic by its code.
+     * @param id - the code of the mechanic to delete
+     **/
+    public void delete(Integer id) {
```
Most of places you are using datatype String for mechanic id and it's fit for this case. 
You can change here also String as datatype for mechanic id.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/ServiceOrderService.java:
```diff
@@ -0,0 +1,91 @@
+    @Transactional(rollbackOn = Exception.class)
+    public ServiceOrder add(ServiceOrderDTO serviceOrder) {
+        try {
+            Set<ServiceOrderLineItem> lineItems = serviceOrder.getLineItems();
```
If you use JoinTable for service order line items then you can persist Service Order & Line Items directly without setting service order id to line items.

Also you have to convert your DTO into entity model class before save.
e.g.
`serviceOrder.setDateRecieved(new Date());`
`return serviceOrderRepository.save(convertToEntity(serviceOrder));`

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/ServiceOrderService.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            serviceOrder.setServiceOrderId(serviceOrderResp.get().getServiceOrderId());
+            serviceOrder.setDateCompleted(new Date());
+            if (!CollectionUtils.isEmpty(serviceOrder.getLineItems())) {
```
If you use JoinTable that time you don't need to apply service order id on line items. I mean it can persist directly but you must have to convert DTO into entity model class.

e.g.
`return serviceOrderRepository.save(convertToEntity(serviceOrder));`

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/ServiceOrderService.java:
```diff
@@ -0,0 +1,91 @@
+            throw new DatabaseErrorException("Failed to delete ServiceOrder", e);
+        }
+    }
+
```
add method to convert dto class to entity model class.

` private ServiceOrder convertToEntity(ServiceOrderDTO serviceOrderDTO) {`
       ` return ServiceOrder.builder().mechanicId(serviceOrderDTO.getMechanicId())`
              `  .customerFirstName(serviceOrderDTO.getCustomerFirstName())`
               ` .customerLastName(serviceOrderDTO.getCustomerLastName())`
               ` .vin(serviceOrderDTO.getVin())`
              `  .dateCompleted(serviceOrderDTO.getDateCompleted())`
              `  .dateRecieved(serviceOrderDTO.getDateRecieved())`
               ` .totalCost(serviceOrderDTO.getTotalCost()).lineItems(serviceOrderDTO.getLineItems()).build();`
    }`

DylanB1205 replied:
So I have made all the changes persisting to the ServiceOrder classes, two things im a little confused on, I have my changes pushed so you can see what im referencing. 1.) IN the ServiceOrder.java the last line I have ServiceType per your suggestions and a created ServiceTyper.java in the model, but im a little lost on if thats right and what goes in that class. 2.) In the ServiceOrderService I think I made the right changes but if you could review them thatd be nice, amd the last convertToEntity doesnt like the .getLineItems() call, should I create a class for it somewhere?

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/ServiceService.java:
```diff
@@ -0,0 +1,39 @@
+package com.wsu.workorderproservice.service;
```
you have incorrect package name here and in import package. Correct that by replacing workorderproservice to shopflowproservice

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/ServiceService.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@RequiredArgsConstructor
```
Missing Service annotation to declare this class as service for Spring Context bean class.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/service/ServiceService.java:
```diff
@@ -0,0 +1,39 @@
+    public List<ServiceDTO> get(boolean isActive) {
+        try {
+            if (isActive) {
+                return mapToDTO(serviceRepository.findByActive(true));
```
you don't have active status in entity model class and table. You can safely delete this method call and remove this method from repository as well. 

You can have findAll method only to return all service types.

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/utilities/Constants.java:
```diff
@@ -0,0 +1,13 @@
+    public static final String DESC = "desc";
+    public static final String PAGE_COUNT = "pageCount";
+    public static final String RESULT_COUNT = "resultCount";
+    public static final String ERROR_MESSAGE = "message";
```
You have same value for this field and MESSAGE field which create duplicate key issue while building meta Map in ControllerExceptionHandler class. Change value for this field as errorMessage

pkumar999 commented on MicroServices/ShopFlowPro-service/src/main/java/com/wsu/shopflowproservice/model/Mechanic.java:
```diff
@@ -0,0 +1,39 @@
+    @Column(name = "mechanic_last_name", nullable = false)
+    private String lastName;
+
+    @Column(name = "specialization", nullable = false)
```
missing this column name in database table.

pkumar999 reviewed (commented):
Great job!!. Your overall changes looks good except few adjustment required to make application fully functional. Which include few DB changes as well to have table & column name in lower case and separated by underscore if it's contains more than two words.

Additionally update **vehicle** and **make** table & it's column as well to have in lower case. And create table for **service_order_line_items** as given below.

CREATE TABLE service_order_line_item (
    service_order_id INT,
    service_id INT,
	PRIMARY KEY (service_order_id, service_id)
);

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated Database

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

TheCodeMaster210 committed:
Finished the Header and edited CSS to ensure it fits properly

TheCodeMaster210 committed:
Added header files

DylanB1205 committed:
Updated

Krackido committed:
adding instructions

Krackido committed:
adding testing information

Krackido committed:
updating test for footer

Krackido committed:
updating header test

Krackido committed:
adding missing sentence

Krackido committed:
adding test

Krackido committed:
more test scripts

Krackido committed:
adding more test

Krackido committed:
correcting script

Krackido committed:
correcting script

Krackido committed:
correcting script again

Krackido committed:
adding more code coverage test

Krackido committed:
adding more test

Krackido committed:
adding test for input change

Krackido committed:
adding test to display vehicles

DylanB1205 committed:
Added components

DylanB1205 committed:
Updated Component

TheCodeMaster210 committed:
fixed margin

TheCodeMaster210 committed:
Merge remote-tracking branch 'origin/main'

DylanB1205 committed:
Updated

TheCodeMaster210 committed:
added comments

TheCodeMaster210 committed:
Merge remote-tracking branch 'origin/main'

TheCodeMaster210 committed:
added comments

DylanB1205 committed:
Updated MicroService

DylanB1205 committed:
Updated MicroService

TheCodeMaster210 committed:
finished

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

Krackido committed:
updating column name

Krackido committed:
Merge branch 'main' of https://github.com/WSU-kduncan/cs3900-autobody-4

Krackido committed:
updating script

Krackido committed:
correcting mechanic

TheCodeMaster210 committed:
Corrected the footer in service-order-details.component.css

TheCodeMaster210 committed:
Merge remote-tracking branch 'origin/main'

Krackido committed:
updating mechanic controller

Krackido committed:
updating script

Krackido committed:
Merge branch 'main' of https://github.com/WSU-kduncan/cs3900-autobody-4

Krackido committed:
correct mechanic data type

Krackido committed:
updating database call

DylanB1205 committed:
Updated DB

DylanB1205 committed:
Updated DB

DylanB1205 committed:
Updated

TheCodeMaster210 committed:
Added a Mechanic profile

TheCodeMaster210 committed:
Added images

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

rjnicolai5 commented:
#Web Design Feedback

## Auto Body:
- Good base functionality. 
- I like the attempts being made here visually. Doesn't need to be much fancier than this
- I'm not sure what exactly is doing it, but when 'ng-serving' and navigating to your site occasionally things will just crash with runtime errors. I'd encourage you to look at your application config files etc. Check your app.routes.ts, anything configuration related that may be getting hung up during run time. The fact your code compiles but the site crashes during runtime is really odd. Try comparing your 'main.ts', 'main.server.ts', 'app.routes.ts' files with our WOMs code maybe. Based on the run time errors occurring I'd be willing to bet it is configuration issues.
	- Alternatively, check that all imports are made correctly. If you use ngIf, ngFor, ngClass, ngStyle, any directive like that, it needs to be imported to the components utilizing it. Check imports as well across the project
- I would encourage moving service files like 'service-order.service' into their own folder. logical organization of the project will really help with giving feedback and diagnosing issues. I.E create a services folder for all services. A model folder for all models. Have all components within the same folder. Organized file structure is very important and can mitigate dependency issues.
- Overall: On the right track, I'd encourage project cleanliness and double checking clean configurations

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

Krackido committed:
attempting get http

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

DylanB1205 committed:
Updated

TheCodeMaster210 committed:
Merge remote-tracking branch 'origin/main'

DylanB1205 committed:
Mechanics GET and DELETE backend to frontend working

TheCodeMaster210 committed:
Merge remote-tracking branch 'origin/main'

TheCodeMaster210 committed:
Removed John-doe files

TheCodeMaster210 committed:
Added Agile requirements to the Project README.md

TheCodeMaster210 committed:
Added Documentation for launching the database to the Project README.md

rjnicolai5 commented:
FE GRADE ESTIMATE: 75

- Big improvements over time across the semester
- Good usage of services, adhering to Single Service Principle, but maybe move service files into their own folder
- Got the footer styled correctly, nice!
- I'd like to see the usage of attribute directives please! ngClass, ngStyle
- Good suite of unit tests, I need to see some more of them passing ideally
- Decent attempts at mimicking what you saw from WOMS, but please remember the http syntax we went over with your project during office hours! I see it commented out in your project. Your http calls won't function fully until this is taken into account
- I don't see any model files! I need to see you guys creating interfaces and data objects that correspond to the data you receive from your service and db

(missing attribute directives, http calls not fully set up, not all unit tests passing, services not organized fully,

DylanB1205 committed:
Mechanics adding, getting, and deleting working frontend to backend

DylanB1205 committed:
Mechanic add, delete and get all working, issues with update

DylanB1205 committed:
Made service-list service files

DylanB1205 committed:
GET services working

TheCodeMaster210 committed:
Update README.md

added checklist section to be filled out later

TheCodeMaster210 committed:
Update README.md

added how to Build front-end program and how for added place holders for building the Backend

TheCodeMaster210 committed:
Update README.md

TheCodeMaster210 committed:
Update README.md

linked Front-end README.MD

TheCodeMaster210 committed:
Added table of contents

TheCodeMaster210 committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	README.md

TheCodeMaster210 committed:
Added table of contents

TheCodeMaster210 committed:
Added table of contents Links

TheCodeMaster210 committed:
Added How to build front end

Krackido committed:
adding testing for service list service

Krackido committed:
Merge branch 'main' of github.com:WSU-kduncan/cs3900-autobody-4 into main

TheCodeMaster210 committed:
Added How to build back end

TheCodeMaster210 committed:
Added Future work Section

TheCodeMaster210 committed:
Added Future work Section

TheCodeMaster210 committed:
Added Future work Section

Krackido committed:
fixing test error

Krackido committed:
removing test error

Krackido committed:
removing button click event

Krackido committed:
fixing test

Krackido committed:
fixing render test

Krackido committed:
fixing test script

Krackido committed:
adding test for template rendering

TheCodeMaster210 committed:
Update README.md to include a list of what's working

TheCodeMaster210 committed:
Update README.md to include Future Work section

TheCodeMaster210 committed:
Update README.md

DylanB1205 committed:
Updated Frontend

DylanB1205 committed:
Merge branch 'main' of github.com:WSU-kduncan/cs3900-autobody-4

DylanB1205 committed:
Updated README

DylanB1205 committed:
Added SwaggerDoc
